File: 백준/Silver/2304. 창고 다각형/창고 다각형.py
n = int(input())
arr = []
max_height = 0
max_idx = 0

# 최대 높이, 최대 높이가 나오는 왼쪽 면의 위치 찾기
for i in range(n):
    l, h = map(int, input().split())
    arr.append((l, h))
    if max_height<h:
        max_height = h
        max_idx_dot = l

# arr 정렬하기
arr.sort(key=lambda x:x[0])

# 좌-우 분기점 찾기
for i in range(len(arr)):
    if arr[i][0]==max_idx_dot:
        max_idx = i
        break


area = 0 # 면적
# 왼쪽
now_max_left = arr[0][1]

for i in range(max_idx):
    if now_max_left<arr[i][1]:
        now_max_left = arr[i][1]
    area += (arr[i+1][0] - arr[i][0])*now_max_left

area += max_height


# 오른쪽
now_max_right = arr[-1][1]

for i in range(n-1, max_idx, -1):
    if now_max_right<arr[i][1]:
        now_max_right = arr[i][1]
    area += (arr[i][0] - arr[i-1][0])*now_max_right

print(area)

File: 백준/Silver/2304. 창고 다각형/창고 다각형.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int maxHeight = 0;
        int maxIdx = 0;
        int maxIdxDot = 0;
        int[][] arr = new int[n][2];
        for (int i = 0; i<n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            arr[i][0] = Integer.parseInt(st.nextToken());
            arr[i][1] = Integer.parseInt(st.nextToken());
            if (maxHeight<arr[i][1]) {
                maxHeight = arr[i][1];
                maxIdxDot = arr[i][0];
            }
        }
        Arrays.sort(arr, (o1, o2) -> {
            return o1[0] - o2[0];
        });

        for (int i = 0; i<n; i++) {
            if (arr[i][0] == maxIdxDot) {
                maxIdx = i;
                break;
            }
        }

        int area = 0;
        int nowMaxLeft = arr[0][1];
        for (int i = 0; i<maxIdx; i++) {
            if (nowMaxLeft<arr[i][1]) {
                nowMaxLeft = arr[i][1];
            }
            area += (arr[i+1][0] - arr[i][0]) * nowMaxLeft;
        }

        area += maxHeight;

        int nowMaxRight = arr[n-1][1];

        for (int i = n-1; i>maxIdx; i--) {
            if (nowMaxRight<arr[i][1]) {
                nowMaxRight = arr[i][1];
            }
            area += (arr[i][0] - arr[i-1][0]) * nowMaxRight;
        }

        System.out.println(area);
    }
}